[cssom] Fix serialization for 'marker' shorthand

Loirooriol opened the pull request:
#### 0a9f425ac8ebf7e8991d438a4b599e4f6c7fe2f1
<pre>
[cssom] Fix serialization for &apos;marker&apos; shorthand
<a href="https://bugs.webkit.org/show_bug.cgi?id=248308">https://bugs.webkit.org/show_bug.cgi?id=248308</a>

Reviewed by Tim Nguyen.

Serializing &apos;marker&apos; at specified-value time would just provide the
value for &apos;marker-start&apos;, without ensuring that &apos;marker-mid&apos; and
&apos;marker-end&apos; were also set to the same value.

Now, if &apos;marker&apos; can&apos;t represent the values for the 3 longhands, it will
serilize as empty string. And then I don&apos;t see the need to disallow
&apos;marker&apos; in cssText.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt:
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::getPropertyValue const):
(WebCore::canUseShorthandForLonghand):

Canonical link: <a href="https://commits.webkit.org/257282@main">https://commits.webkit.org/257282@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ced1f78e45c49c67cdd04b24b4dffe5fe1468600

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98388 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7593 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31517 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107826 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168095 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102326 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8102 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84973 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90949 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/104486 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104043 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6164 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89722 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33160 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87977 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/21084 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/76107 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1541 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/22612 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1475 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45101 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5018 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6383 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42021 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2835 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -1669,8 +1664,6 @@ static constexpr bool canUseShorthandForLonghand(CSSPropertyID shorthandID, CSSP
 
     // FIXME: If font-variant-ligatures is none, this depends on the value of the longhand.
     case CSSPropertyFontVariant:
-    // FIXME: Serialization ignores marker-mid and marker-end (http://webkit.org/b/248308).
-    case CSSPropertyMarker:
```
If we return an empty string for the shorthand, have you checked that all the longhands appear? (I assume yes)

Loirooriol replied:
Yes, `cssText` doesn't use a shorthand whose serialization is a null string, which is the case when `getCommonValue()` doesn't find a common value.

```cpp
            if (!value.isNull()) {
                propertyID = shorthandPropertyID;
                shorthandPropertyUsed.set(shorthandPropertyIndex);
                break;
            }
```

```js
var {style} = document.createElement("div");
style.cssText = 'marker-start: url("a"); marker-mid: url("a"); marker-end: url("a")';
style.marker; // 'url("a")'
style.cssText; // 'marker: url("a");'
style.markerEnd = 'url("b")';
style.marker; // ''
style.cssText; // 'marker-start: url("a"); marker-mid: url("a"); marker-end: url("b");'
```

Loirooriol committed:
[cssom] Fix serialization for 'marker' shorthand
https://bugs.webkit.org/show_bug.cgi?id=248308

Reviewed by Tim Nguyen.

Serializing 'marker' at specified-value time would just provide the
value for 'marker-start', without ensuring that 'marker-mid' and
'marker-end' were also set to the same value.

Now, if 'marker' can't represent the values for the 3 longhands, it will
serilize as empty string. And then I don't see the need to disallow
'marker' in cssText.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt:
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::getPropertyValue const):
(WebCore::canUseShorthandForLonghand):

Canonical link: https://commits.webkit.org/257282@main
